File: backend/src/DB/models/Subscription.ts
import { InferType, number, object, string } from 'yup';
import { privilegesSchema } from './Plan';

export const collectionName = 'subscription'

export const schemaVersion = 'v1.0.0'

export type SubscriptionStatus = 'active' | 'canceled' | 'trial' | 'paymentNotVerified' | 'paymentNotCompleted' | 'inDebtToUser'
const statusSchema = string().oneOf<SubscriptionStatus>(['active', 'canceled', 'trial', 'paymentNotVerified', 'paymentNotCompleted', 'inDebtToUser'])

const processorSubscriptionIdSchema = string().when('status', { is: 'paymentNotVerified', then(s) { return s.optional() }, otherwise(s) { return s.required() } })

export type PaymentMethod = "zarinpal" | "paypal" | "bitcoin" | "zibal"
export type SubscriptionDuration = "month" | "year"
export const paymentMethodSchema = string().oneOf<PaymentMethod>(['zarinpal', 'zibal', 'paypal', 'bitcoin'])
export const subscriptionDurationSchema = string().oneOf<SubscriptionDuration>(['month', 'year'])

export const currencySchema = string().oneOf(['IRR', 'IRT', 'USD', 'EUR', 'BTC', 'ETH'])
export const currencyRequiredSchema = currencySchema.required()
export type Currency = InferType<typeof currencySchema>
export type CurrencyRequired = InferType<typeof currencyRequiredSchema>

export const paymentSchema = object().shape({
    authority: string().optional(),
    completedAt: number().integer().min(0).optional(),
    verifiedAt: number().integer().min(0).optional(),
    refId: string().optional(),
    cardNumber: string().optional(),
    cardNumberHash: string().optional(),
    processorSubscriptionId: processorSubscriptionIdSchema,
    method: paymentMethodSchema.required(),
    currency: currencyRequiredSchema,
    amount: number().integer().min(0).required(), // Negative values indicate we are in dept to user
})
export type Payment = InferType<typeof paymentSchema>

const create = {
    schemaVersion: string().optional().min(6).max(20),
    userId: string().required(),
    planTitle: string().required().label('Plan title'),
    privileges: privilegesSchema.required(),
    status: statusSchema.required(),
    currentPeriodEnd: number().integer().positive().required(),
    payment: paymentSchema.required(),
}
export const subscriptionCreateSchema = object().shape(create).required()

const update = {
    privileges: privilegesSchema.optional(),
    status: statusSchema.optional(),
    currentPeriodEnd: number().optional(),
    payment: paymentSchema.optional(),
}
export const subscriptionUpdateSchema = object().shape(update).required()

export const subscriptionSchema = object().shape({
    schemaVersion: string().optional().min(6).max(20),
    userId: string().required(),
    planTitle: string().required().label('Plan title'),
    privileges: privilegesSchema.required(),
    status: statusSchema.required(),
    currentPeriodEnd: number().integer().positive().required(),
    payment: paymentSchema.required(),
    createdAt: number().optional(),
    updatedAt: number().optional(),
})

export type SubscriptionCreate = InferType<typeof subscriptionCreateSchema>
export type SubscriptionUpdate = InferType<typeof subscriptionUpdateSchema>
export type Subscription = InferType<typeof subscriptionSchema>

File: client/lib/subscription/models/subscription.dart
import 'payment_method.dart';
import 'price.dart';
import 'privileges.dart';
import 'status.dart';

class Subscription {
  final String id;
  final Status status;
  final String planTitle;
  final Price price;
  final Privileges privileges;
  final int currentPeriodEnd;
  final PaymentMethod paymentMethod;

  const Subscription({
    required this.id,
    required this.planTitle,
    required this.price,
    required this.privileges,
    required this.status,
    required this.currentPeriodEnd,
    required this.paymentMethod,
  });

  factory Subscription.fromJson(Map<String, dynamic> json) => Subscription(
    id: json['_id'] as String,
    status: toStatus(json['status'] as String),
    planTitle: json['planTitle'] as String,
    price: Price.fromJson(json['price'] as Map<String, dynamic>),
    privileges: Privileges.fromJson(json['privileges'] as Map<String, dynamic>),
    currentPeriodEnd: json['currentPeriodEnd'] as int,
    paymentMethod: toPaymentMethod(json['paymentMethod'] as String),
  );

  Map<String, dynamic> toJson() => {
    '_id': id,
    'status': status.name,
    'planTitle': planTitle,
    'price': price.toJson(),
    'privileges': privileges.toJson(),
    'currentPeriodEnd': currentPeriodEnd,
    'paymentMethod': paymentMethod.name,
  };
}
